javaslang/src/main/java/javaslang/collection/TreeMap.java

```diff
@@ -477,6 +505,98 @@ public boolean isEmpty() {
         return TreeSet.ofAll(keyComparator(), iterator().map(Tuple2::_1));
     }
 
+    @Override
+    public Iterator<Tuple2<K, V>> descendingIterator() {
+        return entries.descendingIterator();
+    }
+
+    @Override
+    public SortedSet<K> descendingKeySet() {
+        return descendingMap().keySet();
+    }
+
+    @Override
+    public SortedMap<K, V> descendingMap() {
+        return new TreeMap<>(entries.descendingTree());
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(Option<K> fromKey, boolean fromInclusive, Option<K> toKey, boolean toInclusive) {
+        return new TreeMap<>(
+                entries.subTree(
+                    fromKey.map(k -> Tuple.of(k, null)), fromInclusive,
+                    toKey.map(k -> Tuple.of(k, null)), toInclusive));
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(K fromKey, boolean fromInclusive, K toKey, boolean toInclusive) {
+        return subMap(Option.some(fromKey), fromInclusive, Option.some(toKey), toInclusive);
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(K fromKey, K toKey) {
+        return subMap(fromKey, true, toKey, false);
+    }
+
+    @Override
+    public SortedMap<K, V> headMap(K toKey, boolean inclusive) {
+        return subMap(Option.none(), true, Option.some(toKey), inclusive);
+    }
+
+    @Override
+    public SortedMap<K, V> headMap(K toKey) {
+        return headMap(toKey, false);
+    }
+
+    @Override
+    public SortedMap<K, V> tailMap(K fromKey, boolean inclusive) {
+        return subMap(Option.some(fromKey), inclusive, Option.none(), true);
+    }
+
+    @Override
+    public SortedMap<K, V> tailMap(K fromKey) {
+        return tailMap(fromKey, true);
```

Comment:
The whole point of this feature was to provide support for these operations available in `java.util.TreeMap`. The scala `TreeMap` implementation does not support `ceiling/floor/subMap/tailMap/headMap/descendingIterator`. Using your reasoning, we shouldn't implement any of them. 

However, why shouldn't we? Having navigable methods in a sorted set or map is a very important feature in my opinion.

What alternative signature would you suggest for these methods:
* `subMap(K fromKey, boolean fromInclusive, K toKey, boolean toInclusive`
* `subMap(K fromKey, K toKey)`
* `headMap(K toKey, boolean inclusive)`
* `headMap(K toKey)`
* `tailMap(K fromKey, boolean inclusive)`
* `tailMap(K fromKey)`